### PR TITLE
Add performance profiling to flare command

### PR DESCRIPTION
**Pull request by bmikaili**
### What does this PR do?

Adds support for a heap profile and CPU profile to be taken and included in a flare. The duration of the CPU profile can be configured (30s +). Two heap profiles are taken to allow for diff viewing. The Flare process gets the profiles from the expvar endpoint of the agent, encodes them, and sends them over HTTP for the agent to include in the flare. If there's no profile sent, the agent simply doesn't include it.

A prior draft was https://github.com/DataDog/datadog-agent/pull/6016, but this implementation makes less changes for the same outcome so it is preferred.

### Motivation

It would be useful to have performance profiling data for some support cases.

### Describe your test plan

1. Test that flare without the flag doesn't include the profiles.
2. Test that flare returns error when doing profile with a parameter below 30.
3. Test that files are included and `go tool pprof` works on them when profile flag is set with a value above 30.
Extra for QA: Test that flare isn't broken on other paths (windows, GUI).

### CODE REVIEW

File: pkg/flare/archive.go
Comment (ogaca-dd) [inline pkg/flare/archive.go:775]: Missing `defer srcHeapProfile.Close()`

File: pkg/flare/archive.go
Comment (ogaca-dd) [inline pkg/flare/archive.go:804]: There is a`CopyFile` function: https://github.com/DataDog/datadog-agent/blob/18d5c0811f2a9de88f2f48d77eaae19b8c04eea4/pkg/util/common.go#L30.

**Review by ogaca-dd — COMMENTED** [review]
LGTM, just few comments.

File: cmd/agent/api/agent/agent.go
Comment (bmikaili) [reply cmd/agent/api/agent/agent.go:89]: Yeah you are right. I would use an absolute path if I knew how to do it with `mux`.

File: cmd/agent/api/agent/agent.go
Comment (bmikaili) [reply cmd/agent/api/agent/agent.go:86]: I tried doing that, the problem is, that my url looks like that `https://localhost:5001/agent/flare//tmp/e4f171debf982e3e5692025137839` Which I think is messing with the routing

File: cmd/agent/app/flare.go
Comment (bmikaili) [reply cmd/agent/app/flare.go:156]: I see, I didn't know how to pass the path properly in the URL. I will try it with query strings.

File: cmd/agent/app/flare.go
Comment (bmikaili) [reply cmd/agent/app/flare.go:199]: In `writeHTTPCallContent` I use the `os.APPEND` flag so it doesn't override it I think.

File: pkg/flare/archive.go
Comment (bmikaili) [reply pkg/flare/archive.go:111]: I am not sure why this was originally done with a random hex. I am just reusing the function. There might have been a reason to do it like that?

File: cmd/agent/app/flare.go
Comment (bmikaili) [reply cmd/agent/app/flare.go:44]: Will add.

File: pkg/flare/archive.go
Comment (bmikaili) [reply pkg/flare/archive.go:775]: Added

File: pkg/flare/archive.go
Comment (bmikaili) [reply pkg/flare/archive.go:804]: Ah that's nice, didn't know that.

File: cmd/agent/api/agent/agent.go
Comment (ogaca-dd) [inline cmd/agent/api/agent/agent.go:43]: I am wondering if it coud lead to a security issue as this code returns a copy of the files in `profileDir` with the access rights of the running agent.

Comment (bmikaili) [reply cmd/agent/api/agent/agent.go:43]: Technically it just returns the path to the temporary files in `profileDir`, I'm not sure if that's a security issue 🤔

File: cmd/agent/app/flare.go
Comment (gbbr) [inline cmd/agent/app/flare.go:55]: Would it be possible to get the `trace-agent` profiles too? The mechanism and paths are exactly identical, the only thing that would change is the port.

File: cmd/agent/app/flare.go
Comment (gbbr) [inline cmd/agent/app/flare.go:141]: IMHO these "codecov" checks are making reviews quite difficult and are only helpful to the author. They could easily remain as part of the CI.

**Review by gbbr — COMMENTED** [review]
Sorry for coming late to the party, I might miss some context.

File: cmd/agent/app/flare.go
Comment (bmikaili) [reply cmd/agent/app/flare.go:55]: Yeah would definitely be possible. Might have to leave it to a different PR though because I'm not sure I will get to adding that before friday, and I want to merge this before the freeze.

Comment (hush-hush) [inline cmd/agent/app/flare.go:55]: The raw URL should be move to a `const` block along with the one for memory.

File: cmd/agent/app/flare.go
Comment (hush-hush) [inline cmd/agent/app/flare.go:211]: This should be moved to the `flare` package as a helper.

File: cmd/agent/app/flare.go
Comment (hush-hush) [inline cmd/agent/app/flare.go:97]: Should we not use `profiling` somewhere to pull profiling seconds of data ?

File: cmd/agent/app/flare.go
Comment (hush-hush) [inline cmd/agent/app/flare.go:162]: We should stop and exit if we can't serialize the profile. We don't want to send half baked flare.

File: pkg/flare/archive.go
Comment (hush-hush) [inline pkg/flare/archive.go:42]: This is also a `.profile ` and not a log

File: pkg/flare/archive.go
Comment (hush-hush) [inline pkg/flare/archive.go:108]: Why switch argument order ?

File: pkg/flare/archive.go
Comment (hush-hush) [inline pkg/flare/archive.go:789]: Does writing both profile to the same file won't break `pprof` tools ? Should we not write them to 2 different files ?

File: pkg/flare/archive.go
Comment (hush-hush) [inline pkg/flare/archive.go:94]: Should we store a slice of `[]byte` for the heap ? Do we have any use cases where we might want 3 or more heap profile ?

small nit: if we keep the 2 profile could we put them next to each other in the struct:

```go
type Profile struct {
￼	CPUProfile        []byte
￼	FirstHeapProfile  []byte
￼	SecondHeapProfile []byte
}
```

File: cmd/agent/api/agent/agent.go
Comment (hush-hush) [inline cmd/agent/api/agent/agent.go:91]: We should return an error here instead of sending the flare.

File: cmd/agent/api/agent/agent.go
Comment (hush-hush) [inline cmd/agent/api/agent/agent.go:89]: Would it not be simpler to check the size of `r.Body` first instead of starting a new Json decoder ?

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: cmd/agent/app/flare.go
Comment (bmikaili) [reply cmd/agent/app/flare.go:41]: This actually enables both, I just need to mention it in the text.

File: cmd/agent/app/flare.go
Comment (bmikaili) [reply cmd/agent/app/flare.go:103]: The problem with exiting with error is that by default it is set as an invalid time range (0), because profiling is disabled by default. Should I change the default for this?

File: cmd/agent/app/flare.go
Comment (bmikaili) [reply cmd/agent/app/flare.go:55]: Will that update the URL with the value for the profiling seconds though?

File: cmd/agent/app/flare.go
Comment (bmikaili) [reply cmd/agent/app/flare.go:103]: Nvm I'll just make the default -1 lol

File: cmd/agent/app/flare.go
Comment (bmikaili) [reply cmd/agent/app/flare.go:97]: Yes, in the `cpuProfURL`

File: cmd/agent/app/flare.go
Comment (bmikaili) [reply cmd/agent/app/flare.go:55]: If I make them consts, I can't put in the value from the config for the port either with `fmt.Sprintf`

File: pkg/flare/archive.go
Comment (bmikaili) [reply pkg/flare/archive.go:94]: It's only 2 profiles because one is created before the cpu profile, and one after the cpu profile to have a diff. Don't think we need more than that.

Next to each other makes sense 👍

File: cmd/agent/app/flare.go
Comment (bmikaili) [reply cmd/agent/app/flare.go:27]: I see. This happens in a bunch of places though, for example in `archive.go`. I'll refactor those too then?

File: cmd/agent/app/flare.go
Comment (bmikaili) [reply cmd/agent/app/flare.go:55]: Nvm, I moved these to `CreatePerformanceProfile` and moved the method to the `flare` package.

File: pkg/flare/archive.go
Comment (bmikaili) [reply pkg/flare/archive.go:42]: Made it a `.log` because `go-routine-dump` is one as well, but yes, it would be more readable as `.profile`. I will change it.

File: pkg/flare/archive.go
Comment (bmikaili) [reply pkg/flare/archive.go:108]: Because then I only have to write `string` once. It's a minor change but I thought it's nicer.

File: pkg/flare/archive.go
Comment (bmikaili) [reply pkg/flare/archive.go:789]: Yes. Will change together with the file suffix.

File: cmd/agent/api/agent/agent.go
Comment (bmikaili) [reply cmd/agent/api/agent/agent.go:89]: It would be, I did it this way due to the reasoning in [this](https://stackoverflow.com/a/32718077) stackoverflow comment. Do you think we can disregard that as it is just internal communication?

File: cmd/agent/api/agent/agent.go
Comment (bmikaili) [reply cmd/agent/api/agent/agent.go:91]: You mean in the second case?

Comment (hush-hush) [reply cmd/agent/api/agent/agent.go:91]: yes, if the json decode fails

File: cmd/agent/api/agent/agent.go
Comment (hush-hush) [reply cmd/agent/api/agent/agent.go:89]: I was thinking about something like that:

```go
  if r.Body != "" {
	if err := json.NewDecoder(r.Body).Decode(&profile); err != nil {
		// handle error
	}
  }
```

I don't think we need to instantiate and run the JSON decoder if the body is empty.

Comment (bmikaili) [reply cmd/agent/api/agent/agent.go:89]: That makes sense to me, didn't think about doing it that way 👍

File: pkg/flare/archive.go
Comment (truthbk) [reply pkg/flare/archive.go:138]: I agree that it's a nicer interface, it's an unexported function and the likelihood of a regression is small as we're covered by tests and the compiler would fail to build given the change. I'm fine with this.

File: pkg/flare/archive.go
Comment (truthbk) [inline pkg/flare/archive.go:100]: Though this might be out of scope, I feel like it's only right to bring it up. Should we add facilities to collect the relevant profiles from other agents (trace, process,...) exposing the `pprof` API?  It shouldn't be particularly difficult to at least lay the foundation to perform such collection. For instance, we could make the cpuProfURL parametric here:

```golang
// caller is responsible for providing a valid baseURL ie. fmt.Sprintf("http://127.0.0.1:%s/debug/pprof/profile", config.Datadog.GetString("expvar_port")
// but could also be remote, or an alternate port on a different agent.
func CreatePerformanceProfile(baseURL string, profileDuration int) (*Profile, error) {
	cpuProfURL := fmt.Sprintf("%s?seconds=%d", baseURL, profileDuration)
    // ...
}

```

We might also want to consider an alternative where the profiles are submitted via the API in a push model rather than a pull model. Now this is would 100% be out of scope :)

And of course all of the above can be tackled later. So feel free to ignore.

**Review by truthbk — COMMENTED** [review]
Mostly looks good to me after addressing @hush-hush changes, just added a small question so we can maybe reuse some code in future use-cases.

File: pkg/flare/archive.go
Comment (bmikaili) [reply pkg/flare/archive.go:100]: The first one is a quick change and makes sense to me, I'll change it.

The second one would be neat to have, but yeah that one's out of scope. Should I add a card in the backlog or document it elsewhere?

**Review by mx-psi — APPROVED** [review]
Approving; the codeowner request was only because of the shellcheck fix

File: cmd/agent/app/flare.go
Comment (hush-hush) [inline cmd/agent/app/flare.go:158]: Why not simply use `json.Marchall(profile, b)` ?

File: pkg/flare/archive.go
Comment (hush-hush) [inline pkg/flare/archive.go:797]: Let's call `ensureParentDirsExist(cpuProfile)`, just in case,

File: tasks/test.py
Comment (hush-hush) [inline tasks/test.py:595]: Nit: for next time this should be in a different PR.

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: cmd/agent/app/flare.go
Comment (bmikaili) [reply cmd/agent/app/flare.go:158]: In what way? Encode can write to that buffer in less lines.

File: tasks/test.py
Comment (bmikaili) [reply tasks/test.py:595]: It's just cherrypick from master because circleci was failing.

File: pkg/flare/archive.go
Comment (truthbk) [reply pkg/flare/archive.go]: @bmikaili I'll create a card - definitely low priority at the moment.

File: tasks/test.py
Comment (truthbk) [reply tasks/test.py:595]: In this case my advice would be to:
```
# assuming we're in your-branch
git checkout master && git pull
git checkout your-branch
git rebase master
```

or alternatively:
```
# assuming we're in your-branch
git merge origin master
```

File: pkg/flare/archive_test.go
Comment (truthbk) [inline pkg/flare/archive_test.go:32]: We haven't added a test case here, it probably makes sense - you should be able to just use dummy data, even an empty `profile *Profile` to somewhat easily test the functionality.

**Review by truthbk — COMMENTED** [review]
Almost there! Let's add a small test and get this merged tomorrow!

File: tasks/test.py
Comment (hush-hush) [reply tasks/test.py:595]: Yes, `rebase` is your best friend.

File: cmd/agent/api/agent/agent.go
Comment (hush-hush) [inline cmd/agent/api/agent/agent.go:87]: let's use `json.Unmarshal` here too

File: pkg/flare/archive.go
Comment (hush-hush) [inline pkg/flare/archive.go:802]: You opened the file line 797, we already failed if the folder doesn't exists.

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: pkg/flare/archive_test.go
Comment (bmikaili) [reply pkg/flare/archive_test.go:32]: Makes sense to me! I will add tests for `zipPerformanceProfile`. Do you think it also makes sense to unit test the method `CreatePerformanceProfile`? That one does a request against the profile endpoints and creates the `profile` struct from it. I would say that one does not make sense to be tested because it would be completely mocked and it would just be a test of our `apiutil`.

**Review by hush-hush — APPROVED** [review]

**Pull request merged**
